File: gherkin/python/gherkin/ast_node.py
from collections import defaultdict


class AstNode(object):

    def __init__(self, rule_type):
        self.rule_type = rule_type
        self._sub_items = defaultdict(list)

    def add(self, rule_type, obj):
        self._sub_items[rule_type].append(obj)

    def get_single(self, rule_type, default_value=None):
        return self._sub_items[rule_type][0] if self._sub_items[rule_type] else default_value

    def get_items(self, rule_type):
        return self._sub_items[rule_type]

    def get_token(self, token_type):
        return self.get_single(token_type)

    def get_tokens(self, token_type):
        return self._sub_items[token_type]

File: gherkin/java/src/main/java/gherkin/AstNode.java
package gherkin;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static gherkin.Parser.RuleType;
import static gherkin.Parser.TokenType;

public class AstNode {
    private final Map<RuleType, List<Object>> subItems = new HashMap<RuleType, List<Object>>();
    public final RuleType ruleType;

    public AstNode(RuleType ruleType) {
        this.ruleType = ruleType;
    }

    public void add(RuleType ruleType, Object obj) {
        List<Object> items = subItems.get(ruleType);
        if (items == null) {
            items = new ArrayList<Object>();
            subItems.put(ruleType, items);
        }
        items.add(obj);
    }

    public <T> T getSingle(RuleType ruleType, T defaultResult) {
        List<Object> items = getItems(ruleType);
        return (T) (items.isEmpty() ? defaultResult : items.get(0));
    }

    public <T> List<T> getItems(RuleType ruleType) {
        List<T> items = (List<T>) subItems.get(ruleType);
        if (items == null) {
            return Collections.emptyList();
        }
        return items;
    }

    public Token getToken(TokenType tokenType) {
        RuleType ruleType = RuleType.cast(tokenType);
        return getSingle(ruleType, new Token(null, null));
    }

    public List<Token> getTokens(TokenType tokenType) {
        return getItems(RuleType.cast(tokenType));
    }
}
